### ISSUE TITLE
JS error when uploading avatar or overwriting content-image.

### ISSUE THREAD

**Issue by dmotion** [getkirby/panel#570]
When i try to upload an avatar, I got an JS error.

**Uncaught TypeError: fail is not a function - panel.js #758**

Although it's funny that failure is not an option, this bug prevends me from uploading an image to my account.

Tested it (to be sure) on several servers (local- and remote) and browsers (all Windows 10 - but both Chrome, IE and FF) and the error still occurs.

It starts with a 400-server error (bad request) and end's up with a JS error (see screenshot).

![clipboard-1](https://cloud.githubusercontent.com/assets/2298150/10462888/6c722efe-71e1-11e5-9b84-ac2e3fd5113a.png)

**local host**

![image](https://cloud.githubusercontent.com/assets/2298150/10462958/e42a4198-71e1-11e5-93ee-d7d010ee3764.png)

**remote host**

**Comment by texnixe** [comment]
Have you tried to re-install/re-upload the panel folder?

**Comment by dmotion** [comment]
Just did - I downloaded the Panel-Master from Git... still got the error, like the one above.

The POST error did throw an URI to the uploaded avatar... so I followed the link in the console-panel and that resulted in a kind of JSON-string;

**{"status":"error","code":400,"message":"Invalid API method","data":[]}**

![image](https://cloud.githubusercontent.com/assets/2298150/10480457/9ae21874-726a-11e5-8f35-7da950700cd2.png)

**Comment by dmotion** [comment]
As a work-around, is there a way to manually attach an avatar to your profile (dragging / dropping an icon in the avatar-folder and adjusting your profile-code somewhere?).

**Comment by distantnative** [comment]
You just need to put a file in the avatar folder that has the username as its name (allowed formats: `'jpg', 'jpeg', 'png', 'gif'`)

**Comment by dmotion** [comment]
![image](https://cloud.githubusercontent.com/assets/2298150/10484338/d70daaba-7284-11e5-8a45-cab761be5f5f.png)

Thanks for the tip - I knew it was that easy, 'cause it's in the philosophy of how Kirby works.

The bug (maybe it's just mine...) is still there, but at least I can see my own avatar in the panel now.

**Comment by dmotion** [comment]
Okay - here I am again...

It seems I can not only NOT upload images for the avatar, but also not REPLACING an existing image in the default panel.

I can upload a picture once, but when I try to replace that image (by the panel) I always got the error **The uploaded file must have the same file type**

![image](https://cloud.githubusercontent.com/assets/2298150/10486274/866af9a0-728e-11e5-9ef7-181ae70f2994.png)

Guess what... the image I try to upload is 100% the same as the original (same size / format / extension)... and even than it's not working.

When I open the console, I see the same error as always...

![image](https://cloud.githubusercontent.com/assets/2298150/10486304/aa5a9a64-728e-11e5-8cd8-ba8bcbec506f.png)

And when I follow one of the links in the console, I got this message;

**{"status":"error","code":400,"message":"Invalid API method","data":[]}**

![image](https://cloud.githubusercontent.com/assets/2298150/10486325/c1330fa0-728e-11e5-88d4-75ceff033ffc.png)

It occurs on both a local-server (XAMPP) and remote, with several browsers.

It this a bug, or just me?

**Comment by dmotion** [comment]
<h3>Work around, with code fix</h3>

Okay, I created a dirty solution myself.

1). Open the file **/panel/app/controllers/api/files.php**
2). Select all the code in the function **public function upload()** at line #5
3). Copy this code to the clipboard
4). Paste the "upload code" in the function **public function replace()** at line #55

Now you can both **upload** and **rewrite** files, even on shared servers (be sure the name of the new file is the same as the old file).

Not very gentle, but I can use the panel now...

**Comment by dmotion** [comment]
<h3>Update - even better fix</h3>

When using the code below in the files-api, you can even upload a file with a complete different name.

The new file is renamed after the original one, so you don't have to bother for that.

```php
  public function replace($id = null) {

    if(!get('_csrf') or !csrf(get('_csrf'))) {
      return response::error('unauthenticated access');
    }

    $page      = $this->page($id);
    $blueprint = blueprint::find($page);
    $filename  = get('filename');

    $upload = new Upload($page->root() . DS . $filename, array(
      'overwrite' => true,
      'accept'    => function($file) {

        $callback = kirby()->option('panel.upload.accept');

        if(is_callable($callback)) {
          return call($callback, $file);
        } else {
          return true;
        }

      }
    ));

    if($file = $upload->file()) {
      try {
        $this->checkUpload($file, $blueprint);
        kirby()->trigger('panel.file.replace', $file);
        return response::success('success');
      } catch(Exception $e) {
        $file->delete();
        return response::error($e->getMessage());
      }
    } else {
      return response::error($upload->error()->getMessage());
    }

  }

```

Tested and works on all server, local / remote / shared and dedicated.

**Comment by bastianallgeier** [comment]
This should be fixed in 2.2. The update is already available on the Decruft branch, but beta yet.
